### PR TITLE
Graeme/dls 8523/utr page

### PR DESCRIPTION
**Pull request by gjstirling**

### CODE REVIEW

File: conf/messages [hunk 1]
```diff
@@ -718,6 +720,9 @@ fh.businessPartners.partnershipName.error.pattern=Partnership name must be 120 c
 fh.business_partners.utr.link1 = You can
 fh.business_partners.utr.link2 = find a lost UTR.
 
+fh.businessPartners.SoleProprietor.utr.title=What is the partner’s Corporation Tax Unique Taxpayer Reference (UTR)?
+fh.businessPartners.SoleProprietor.utr.heading=What is {0}’s Self Assessment Unique Taxpayer Reference (UTR)?
```
Comment (BackRowOla) [inline conf/messages:724]: Duplicated message - already added on line 694

File: app/uk/gov/hmrc/fhregistrationfrontend/views/businessPartners/v2/business_partners_SoleProprietors_utr.scala.html [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    Messages("fh.businessPartners.SoleProprietor.utr.title")
+}
+
+@heading = @{
+    Messages("fh.businessPartners.SoleProprietor.utr.heading", partnerName)
+}
+
+@layout(title = titlePrefix + title, backUrl = Some("#")) {
+
+    @Html(s"${BusinessPartnersHeader(heading)}")
```
Comment (BackRowOla) [inline app/uk/gov/hmrc/fhregistrationfrontend/views/businessPartners/v2/business_partners_SoleProprietors_utr.scala.html:47]: String interpolation not needed - @Html(BusinessPartnersHeader(heading)) should work.

File: app/uk/gov/hmrc/fhregistrationfrontend/views/businessPartners/v2/business_partners_SoleProprietors_utr.scala.html [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            @viewHelpers.govukErrorSummary(ErrorSummary(errorList =
+                Seq(ErrorLink(
+                    href = Some(s"#${businessPartnersUtrForm.errors.head.key}"),
+                    content = Text(messages(s"fh.${businessPartnersUtrForm.errors.head.key}.${businessPartnersUtrForm.errors.head.message}"))
+                )),
+                title = Text(messages("fh.generic.error"))
+            ))
+        }
+
+        @Html(s"${LostUtrLink()}")
```
Comment (BackRowOla) [inline app/uk/gov/hmrc/fhregistrationfrontend/views/businessPartners/v2/business_partners_SoleProprietors_utr.scala.html:60]: Same as above - string interpolation not needed

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala [hunk 4]
```diff
@@ -5,64 +5,64 @@ import play.api.libs.ws.DefaultWSCookie
...
 import uk.gov.hmrc.fhregistrationfrontend.testsupport.{Specifications, TestConfiguration}
 
-class BusinessPartnersCorporateBodyUtrControllerIntegrationSpec
+class BusinessPartnersSoleProprietorUtrControllerISpec
   extends Specifications with TestConfiguration {
 
-  "GET /form/business-partners/corporate-body-corporation-tax-unique-taxpayer-reference" should {
+  val route = "/business-partners/partner-self-assessment-unique-taxpayer-reference"
+
+  s"GET /business-partners/partner-self-assessment-unique-taxpayer-reference" should {
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala:13]: could use the route here like you've done in the CorporateBodyISpec

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala [hunk 5]
```diff
@@ -5,64 +5,64 @@ import play.api.libs.ws.DefaultWSCookie
...
-          page.getElementsByTag("h1").text() must include("Does test partner have a Self Assessment Unique Taxpayer Reference (UTR)?")
+          page.title() must include("What is the partner’s Corporation Tax Unique Taxpayer Reference (UTR)?")
+          page.getElementsByTag("h1").text() must include("What is {{partner name}}’s Self Assessment Unique Taxpayer Reference (UTR)?")
         }
       }
     }
   }
 
-  "POST /form/business-partners/corporate-body-corporation-tax-unique-taxpayer-reference" when {
+  "POST /business-partners/partner-self-assessment-unique-taxpayer-reference" when {
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala:34]: Same as route comment above - could use the route here like you've done in the CorporateBodyISpec + missing prefix

File: conf/app.routes [hunk 6]
```diff
@@ -130,6 +130,9 @@ POST         /form/business-partners/partnership-name
 GET          /form/business-partners/confirm-corporate-body-registered-office-address   @uk.gov.hmrc.fhregistrationfrontend.controllers.BusinessPartnersConfirmCorporateRegisteredAddressController.load()
 POST         /form/business-partners/confirm-corporate-body-registered-office-address   @uk.gov.hmrc.fhregistrationfrontend.controllers.BusinessPartnersConfirmCorporateRegisteredAddressController.next()
 
+GET          /business-partners/partner-self-assessment-unique-taxpayer-reference       @uk.gov.hmrc.fhregistrationfrontend.controllers.BusinessPartnerSoleProprietorUtrController.load()
+POST         /business-partners/partner-self-assessment-unique-taxpayer-reference       @uk.gov.hmrc.fhregistrationfrontend.controllers.BusinessPartnerSoleProprietorUtrController.next()
```
Comment (BackRowOla) [inline conf/app.routes:134]: Missing /form/... prefix from the route

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          whenReady(result) { res =>
+            res.status mustBe 200
+            res.body mustBe "Next page! with UTR: BusinessPartnersEnterUniqueTaxpayerReference(1234567890)"
+          }
+        }
+      }
+    }
+
+
+    "option selected but UTR not supplied" should {
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala:56]: No option being selected. Please change to remove reference to selecting an option.

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala (hunk 4 above)
Comment (BackRowOla) [reply it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala:13]: Also its missing the /form/... prefix

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  val route = "/form/business-partners/corporate-body-corporation-tax-unique-taxpayer-reference"
+
+  s"GET $route" should {
+
+    "render the corporate-body-corporation-tax-unique-taxpayer-reference page" in {
+      given
+        .commonPrecondition
+
+      WsTestClient.withClient { client =>
+        val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala:20]: Why use both string interpolation and concatenation? You could use 
`val result = client.url(baseUrl + route)` or `val result = client.url(s"$baseUrl$route")`
Doing both just adds extra steps.

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  s"POST $route" when {
+    "The UTR is entered" should {
+      "return 200 with UTR" in {
+        given
+          .commonPrecondition
+
+        WsTestClient.withClient { client =>
+          val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala:41]: See string comment above

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+
+    "option selected but UTR not supplied" should {
+      "return 400" in {
+        given
+          .commonPrecondition
+
+        WsTestClient.withClient { client =>
+          val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala:62]: See string comment above

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      }
+    }
+
+    "the UTR field is invalid" should {
+      "return 400" in {
+        given
+          .commonPrecondition
+
+        WsTestClient.withClient { client =>
+          val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersCorporateBodyUtrControllerISpec.scala:83]: See string comment above

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala [hunk 12]
```diff
@@ -5,64 +5,64 @@ import play.api.libs.ws.DefaultWSCookie
...
+
+  s"GET /business-partners/partner-self-assessment-unique-taxpayer-reference" should {
 
     "render the partnership-self-assessment-unique-taxpayer-reference page" in {
       given
         .commonPrecondition
 
       WsTestClient.withClient { client =>
-        val result = client.url(s"$baseUrl/form/business-partners/partnership-self-assessment-unique-taxpayer-reference")
+        val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala:20]: See string comment above - from comments on it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala [hunk 13]
```diff
@@ -5,64 +5,64 @@ import play.api.libs.ws.DefaultWSCookie
...
-  "POST /form/business-partners/corporate-body-corporation-tax-unique-taxpayer-reference" when {
+  "POST /business-partners/partner-self-assessment-unique-taxpayer-reference" when {
     "yes is selected and the UTR is entered" should {
       "return 200 with UTR" in {
         given
           .commonPrecondition
 
         WsTestClient.withClient { client =>
-          val result = client.url(s"$baseUrl/form/business-partners/corporate-body-corporation-tax-unique-taxpayer-reference")
+          val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala:41]: See string comment above

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala [hunk 14]
```diff
@@ -5,64 +5,64 @@ import play.api.libs.ws.DefaultWSCookie
...
-
-    "option selected but UTR not supplied" should {
+    "No Utr number is supplied" should {
       "return 400" in {
         given
           .commonPrecondition
 
         WsTestClient.withClient { client =>
-          val result = client.url(s"$baseUrl/form/business-partners/corporate-body-corporation-tax-unique-taxpayer-reference")
+          val result = client.url(s"$baseUrl" + s"$route")
```
Comment (BackRowOla) [inline it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala:61]: See string comment above

**Review by BackRowOla — CHANGES_REQUESTED** [review]
Looks good for the most part - a few changes to be made.

File: it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala (hunk 4 above)
Comment (gjstirling) [reply it/uk/gov/hmrc/fhregistrationfrontend/controllers/BusinessPartnersSoleProprietorUtrControllerISpec.scala]: form prefix not required, confirmed by James/Ola.

File: app/uk/gov/hmrc/fhregistrationfrontend/forms/models/hasUniqueTaxpayerReference.scala [hunk 15]
```diff
@@ -18,11 +18,11 @@ package uk.gov.hmrc.fhregistrationfrontend.forms.models
 
 import play.api.libs.json.Json
 
-case class UniqueTaxpayerReference(
+case class hasUniqueTaxpayerReference(
```
[the remark this answers is not in the document]
Comment (gjstirling) [reply app/uk/gov/hmrc/fhregistrationfrontend/forms/models/hasUniqueTaxpayerReference.scala]: Done

File: app/uk/gov/hmrc/fhregistrationfrontend/forms/models/hasUniqueTaxpayerReference.scala [hunk 16]
```diff
@@ -18,11 +18,11 @@ package uk.gov.hmrc.fhregistrationfrontend.forms.models
...
   hasValue: Boolean,
   value: Option[String]
 )
 
-object UniqueTaxpayerReference {
-  implicit val format = Json.format[UniqueTaxpayerReference]
+object hasUniqueTaxpayerReference {
+  implicit val format = Json.format[hasUniqueTaxpayerReference]
```
[the remark this answers is not in the document]
Comment (gjstirling) [reply app/uk/gov/hmrc/fhregistrationfrontend/forms/models/hasUniqueTaxpayerReference.scala]: Done

File: conf/messages (hunk 1 above)
Comment (gjstirling) [reply conf/messages:722]: removed line 694

File: app/uk/gov/hmrc/fhregistrationfrontend/views/businessPartners/v2/business_partners_SoleProprietors_utr.scala.html (hunk 2 above)
Comment (gjstirling) [reply app/uk/gov/hmrc/fhregistrationfrontend/views/businessPartners/v2/business_partners_SoleProprietors_utr.scala.html]: Removed string interpolation

**Review by jackwillett1 — APPROVED** [review]

**Pull request merged**